libnavigation-core/src/test/java/com/mapbox/navigation/core/accounts/BillingControllerWaypointsComparisonTest.kt

```diff
@@ -0,0 +1,517 @@
+package com.mapbox.navigation.core.accounts
+
+import com.mapbox.common.BillingServiceInterface
+import com.mapbox.common.BillingSessionStatus
+import com.mapbox.common.SessionSKUIdentifier
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.arrival.ArrivalProgressObserver
+import com.mapbox.navigation.core.trip.session.NavigationSession
+import com.mapbox.navigation.core.trip.session.NavigationSessionStateObserver
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.testing.LoggingFrontendTestRule
+import com.mapbox.navigation.testing.factories.createDirectionsRoute
+import com.mapbox.navigation.testing.factories.createNavigationRoute
+import com.mapbox.navigator.Waypoint
+import com.mapbox.navigator.WaypointType
+import io.mockk.Runs
+import io.mockk.every
+import io.mockk.just
+import io.mockk.mockk
+import io.mockk.mockkObject
+import io.mockk.slot
+import io.mockk.verify
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+@RunWith(Parameterized::class)
+class BillingControllerWaypointsComparisonTest(
+    private val description: String,
+    private val remainingWaypoints: Int,
+    private val oldWaypoints: List<Waypoint>,
+    private val newWaypoints: List<Waypoint>,
+    private val shouldStartSession: Boolean,
+) {
+
+    companion object {
+
+        @JvmStatic
+        @Parameterized.Parameters(name = "{0}")
+        fun data(): Collection<Array<Any>> {
+            return listOf(
+                arrayOf(
+                    "Route start, non EV + non EV with non-matching size",
+                    2,
+                    listOf(
+                        waypoint(2.2, 3.3, WaypointType.REGULAR),
+                        waypoint(3.3, 4.4, WaypointType.REGULAR),
+                        waypoint(4.4, 5.5, WaypointType.REGULAR)
+                    ),
+                    listOf(
+                        waypoint(1.1, 2.2, WaypointType.REGULAR),
+                        waypoint(3.3, 4.4, WaypointType.REGULAR),
+                    ),
+                    true
+                ),
```

Comment:
While this and some tests below mention that they check for the mismatch in the number of waypoints, they would always pass even if the number was the same because the destination points are different. It'd be good to align the test cases depending on what they should actually test, so that there's only one failure/success point per test.